use-cases/sports-league-scheduling/src/main/java/org/acme/sportsleagueschedule/solver/SportsLeagueSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,127 @@
+package org.acme.sportsleagueschedule.solver;
+
+import static ai.timefold.solver.core.api.score.stream.Joiners.equal;
+import static ai.timefold.solver.core.api.score.stream.Joiners.filtering;
+
+import ai.timefold.solver.core.api.score.buildin.hardsoft.HardSoftScore;
+import ai.timefold.solver.core.api.score.stream.Constraint;
+import ai.timefold.solver.core.api.score.stream.ConstraintFactory;
+import ai.timefold.solver.core.api.score.stream.ConstraintProvider;
+
+import org.acme.sportsleagueschedule.domain.Match;
+import org.acme.sportsleagueschedule.domain.Round;
+
+public class SportsLeagueSchedulingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                matchesOnSameDay(constraintFactory),
+                fourConsecutiveHomeMatches(constraintFactory),
+                fourConsecutiveAwayMatches(constraintFactory),
+                repeatMatchOnTheNextDay(constraintFactory),
+                startToAwayHop(constraintFactory),
+                homeToAwayHop(constraintFactory),
+                awayToAwayHop(constraintFactory),
+                awayToHomeHop(constraintFactory),
+                awayToEndHop(constraintFactory),
+                classicMatches(constraintFactory)
+        };
+    }
+
+    protected Constraint matchesOnSameDay(ConstraintFactory constraintFactory) {
+        return constraintFactory
+                .forEachUniquePair(Match.class,
+                        equal(Match::getRoundIndex),
+                        filtering((match1, match2) -> match1.getHomeTeam().equals(match2.getHomeTeam())
+                                || match1.getHomeTeam().equals(match2.getAwayTeam())
+                                || match1.getAwayTeam().equals(match2.getAwayTeam())))
+                .penalize(HardSoftScore.ONE_HARD)
+                .asConstraint("matches on the same day");
+    }
+
+    protected Constraint fourConsecutiveHomeMatches(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEach(Match.class)
+                .ifExists(Match.class, equal(Match::getHomeTeam),
+                        equal(match -> match.getRoundIndex() + 1, Match::getRoundIndex))
+                .ifExists(Match.class, equal(Match::getHomeTeam),
+                        equal(match -> match.getRoundIndex() + 2, Match::getRoundIndex))
+                .ifExists(Match.class, equal(Match::getHomeTeam),
+                        equal(match -> match.getRoundIndex() + 3, Match::getRoundIndex))
```

Comment:
Consider using the consecutive collector for this. (See either docs or nurse rostering.) Otherwise we really don't have an example to showcase it anymore.

Reply:
Nice catch!